Firestore/core/src/firebase/firestore/remote/stream.h

```diff
@@ -0,0 +1,246 @@
+   */
+  bool IsOpen() const;
+
+  /**
+   * After an error, the stream will usually back off on the next attempt to
+   * start it. If the error warrants an immediate restart of the stream, the
+   * sender can use this to indicate that the receiver should not back off.
+   *
+   * Each error will call `OnStreamClose`. That function can decide to
+   * cancel backoff if required.
+   */
+  void InhibitBackoff();
+
+  /**
+   * Marks this stream as idle. If no further actions are performed on the
+   * stream for one minute, the stream will automatically close itself and
+   * notify the stream's `OnClose` handler with Status::OK. The stream will then
+   * be in a non-started state, requiring the caller to start the stream again
+   * before further use.
+   *
+   * Only streams that are in state 'Open' can be marked idle, as all other
+   * states imply pending network operations.
+   */
+  void MarkIdle();
+
+  /**
+   * Marks the stream as active again, preventing auto-closing of the stream.
+   * Can be called from any state -- if the stream is not in state `Open`, this
+   * is a no-op.
+   */
+  void CancelIdleCheck();
+
+  // `GrpcStreamObserver` interface -- do not use.
+  void OnStreamStart() override;
+  void OnStreamRead(const grpc::ByteBuffer& message) override;
+  void OnStreamError(const util::Status& status) override;
+
+ protected:
+  // `Stream` expects all its methods to be called on the worker queue.
+  void EnsureOnQueue() const;
+  void Write(grpc::ByteBuffer&& message);
+  void ResetBackoff();
+  std::string GetDebugDescription() const;
+
+ private:
+  // The interface for the derived classes.
+
+  virtual std::unique_ptr<GrpcStream> CreateGrpcStream(
+      Datastore* datastore, absl::string_view token) = 0;
+  // PORTING NOTE: equivalent to `tearDown`.
+  virtual void FinishGrpcStream(GrpcStream* stream) = 0;
+  virtual void DoOnStreamStart() = 0;
+  virtual util::Status DoOnStreamRead(const grpc::ByteBuffer& message) = 0;
+  virtual void DoOnStreamFinish(const util::Status& status) = 0;
+  // PORTING NOTE: C++ cannot rely on RTTI, unlike other platforms.
+  virtual std::string GetDebugName() const = 0;
+
+  // Used to prevent auth if the stream happens to be restarted before token is
+  // received.
+  void RaiseGeneration() {
+    ++generation_;
+  }
+
+  void Authenticate();
+  void ResumeStartAfterAuth(const util::StatusOr<auth::Token>& maybe_token);
+
+  void BackoffAndTryRestarting();
+  void ResumeStartFromBackoff();
+  void StopDueToIdleness();
+
+  void ResetGrpcStream();
+
+  State state_ = State::Initial;
+
+  std::unique_ptr<GrpcStream> grpc_stream_;
+
+  auth::CredentialsProvider* credentials_provider_ = nullptr;
+  util::AsyncQueue* worker_queue_ = nullptr;
+  Datastore* datastore_ = nullptr;
+
+  ExponentialBackoff backoff_;
+  util::TimerId idle_timer_id_{};
+  util::DelayedOperation idleness_timer_;
+
+  // Generation is incremented in each call to `Stop`.
+  int generation_ = 0;
```

Comment:
Slight preference because "close count" seemed more explicit (I can guess exactly what it's trying by the name, but "generation" is a lot more vague).

Reply:
Changed to `close_count_`.

I actually prefer "generation" to "close count", because in my view "close count" is an implementation detail, whereas "generation" is self-documenting:
```cpp
void Stream::Auth(int auth_close_count) {
  if (close_count_ > auth_close_count) {
    return; // What does close count have to do with auth?
  }
  if (generation_ > auth_generation) {
    return; // I don't know when generation is increased, but I can understand the intention here
  }
```